pkg/util/network.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"net"
+	"regexp"
+)
+
+// hostNameFormat matches strings that:
+//
+//   - do not exceed 63 characters
+//   - contain only alpha-numeric characters, dashes, or symbol-point unicode
+//     (ex. emoji)
+//   - does not begin or end with a dash
+const hostNameFormat = `^([a-zA-Z0-9\p{S}\p{L}]{1,2}|(?:[a-zA-Z0-9\p{S}\p{L}][a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}]))$`
+
+var hostNameRx = regexp.MustCompile(hostNameFormat)
+
+// IsValidHostName returns true if the provided value adheres to the format of
+// an RFC-1123 DNS label, which may be:
+//
+//   - an IP4 address
+//   - an IP6 address
+//   - a string of alpha-numeric characters that:
+//   - must not exceed 63 characters in length
+//   - may begin with a digit
+//   - may contain dashes except as the first or last character
+//   - must not contain underscores
+//   - may contain symbol-point unicode, (ex. emoji)
+func IsValidHostName(s string) bool {
```

Comment:
Can we use https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L187?

Reply:
Nope. I spent far too much time looking into this and whether to use:

* https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go
* or https://github.com/kubernetes/kubernetes/blob/ac8ac0fc170eec7cdf292d841f8d026a65ce1681/vendor/github.com/go-openapi/strfmt/default.go
* which is based on https://github.com/go-openapi/strfmt/blob/master/default.go

I read both [RFC-1123](https://datatracker.ietf.org/doc/html/rfc1123#page-13) and [RFC-1034, Section 3.5](https://datatracker.ietf.org/doc/html/rfc1034#section-3.5), and ultimately the Kubernetes validation is very specific to DNS names as labels, not DNS names as host and/or domain names. The code from https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go:

* requires a minimum of two segments for validating a domain name
* does not allow:
   * upper-case
   * unicode
* does not restrict length specifically for Windows guests
* IP4 or IP6 addresses as the host name

The regular expressions I crafted are based on [`HostNamePattern`](https://github.com/go-openapi/strfmt/blob/f065ed8a46434a6407610166cd707c170cf1e483/default.go#L33-L60) from OpenAPI:

```golang
	// HostnamePattern http://json-schema.org/latest/json-schema-validation.html#anchor114
	//  A string instance is valid against this attribute if it is a valid
	//  representation for an Internet host name, as defined by RFC 1034, section 3.1 [RFC1034].
	//  http://tools.ietf.org/html/rfc1034#section-3.5
	//  <digit> ::= any one of the ten digits 0 through 9
	//  var digit = /[0-9]/;
	//  <letter> ::= any one of the 52 alphabetic characters A through Z in upper case and a through z in lower case
	//  var letter = /[a-zA-Z]/;
	//  <let-dig> ::= <letter> | <digit>
	//  var letDig = /[0-9a-zA-Z]/;
	//  <let-dig-hyp> ::= <let-dig> | "-"
	//  var letDigHyp = /[-0-9a-zA-Z]/;
	//  <ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
	//  var ldhStr = /[-0-9a-zA-Z]+/;
	//  <label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]
	//  var label = /[a-zA-Z](([-0-9a-zA-Z]+)?[0-9a-zA-Z])?/;
	//  <subdomain> ::= <label> | <subdomain> "." <label>
	//  var subdomain = /^[a-zA-Z](([-0-9a-zA-Z]+)?[0-9a-zA-Z])?(\.[a-zA-Z](([-0-9a-zA-Z]+)?[0-9a-zA-Z])?)*$/;
	//  <domain> ::= <subdomain> | " "
	//
	// Additional validations:
	//   - for FDQNs, top-level domain (e.g. ".com"), is at least to letters long (no special characters here)
	//   - hostnames may start with a digit [RFC1123]
	//   - special registered names with an underscore ('_') are not allowed in this context
	//   - dashes are permitted, but not at the start or the end of a segment
	//   - long top-level domain names (e.g. example.london) are permitted
	//   - symbol unicode points are permitted (e.g. emoji) (not for top-level domain)
	HostnamePattern = `^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z\p{L}]){2,63})$`
```

Ultimately I took the above and crafted two patterns, one that restricts it to a single segment (host name) and one that allows for an FQDN (domain name). FWIW, the tests I wrote far exceed anything in the referenced K8s or OpenAPI repos :)

Reply:
Thanks for the detailed reply.  Resolving. :)